File: eric/gnyr23/forestforthetrees.py
# all possible directions, U L D R
directions = {
	"R": (lambda x, y: (x, y)),
	"U": (lambda x, y: (y, -x)),
	"L": (lambda x, y: (-x, -y)),
	"D": (lambda x, y: (-y, x))
}


def solve():
	N_TREES, N_SENSORS, R_MAX = map(int, input().strip().split())
	trees = set()
	for _ in range(N_TREES):
		x, y = map(int, input().strip().split())
		trees.add((x, y))
	sensors = []
	for _ in range(N_SENSORS):
		s_x, s_y = map(int, input().strip().split())
		sensors.append((s_x, s_y))  # sensor positions

	found = None  # valid positions go here

	def check(p_x: int, p_y: int, key: str, tree: tuple[int, int]):
		# validate position
		mapped_trees = {tree}
		adj = directions[key]
		for sensor in sensors[1:]:
			s_x, s_y = sensor
			adj_x, adj_y = adj(s_x, s_y)
			t_x = p_x + adj_x
			t_y = p_y + adj_y
			target_tree = (t_x, t_y)
			# is there a tree at this sensors position
			if not target_tree in trees:
				return False
			mapped_trees.add(target_tree)

		if not len(mapped_trees) == N_SENSORS:  # Unable to map one tree to each sensor
			return False

		# ensure any unmapped trees are not within R_MAX
		for t in set(trees - mapped_trees):
			t_x, t_y = t
			m_dist = abs(t_x - p_x) + abs(t_y - p_y)
			if m_dist > R_MAX:  # tree outside robot's range, does not need to be mapped
				continue
			return False  # "detectable" tree within robots range, invalid position
		return True

	for k, direction in directions.items():
		for t in trees:
			t_x, t_y = t
			s_x, s_y = sensors[0]  # all sensors will have to be mapped
			adj_x, adj_y = direction(s_x, s_y)
			x = t_x - adj_x
			y = t_y - adj_y
			pos = (x, y)  # calculate possible position using any tree, sensor, and direction
			if pos in trees:  # robot cannot be on a tree
				continue
			if not check(x, y, k, t):  # check if the robot is possibly at this position
				continue
			if found:  # Location already found, we just found another
				return "Ambiguous"
			found = (x, y)
	if not found:  # no valid positions found
		return "Impossible"
	return " ".join(map(str, found))


print(solve())

File: eric/gnyr23/forestforthetrees.cpp
#include <bits/stdc++.h>
using namespace std;

enum direction { R = 'R', U = 'U', L = 'L', D = 'D' };
typedef pair<int, int> p;
int n_t;
int n_s;
int r_max;
set<p> trees;
vector<p> sensors;
p found;
p *is_found = &found;

p adj(p o_p, direction d) {
    int x = o_p.first;
    int y = o_p.second;
    p n_p;
    switch (d) {
        case R:
            n_p.first = x;
            n_p.second = y;
            break;
        case U:
            n_p.first = y;
            n_p.second = -1 * x;
            break;
        case L:
            n_p.first = -1 * x;
            n_p.second = -1 * y;
            break;
        case D:
            n_p.first = -1 * y;
            n_p.second = x;
            break;
    }
    return n_p;
}

bool check(p pos, p t, direction d) {
    set<p> mapped;
    mapped.insert(t);
    for (int i = 1; i < sizeof sensors; i++) {
        p a = adj(sensors[i], d);
        p target = pair(pos.first + a.first, pos.second + a.second); // position to check
        if (trees.find(target) != trees.end()) { // no t at pos
            return false;
        }
        mapped.insert(target);
    }
    if (mapped.size() != n_s) { // not all sensors mapped
        return false;
    }
    vector<p> unmapped;
    set_difference(trees.begin(), trees.end(),
                   mapped.begin(), mapped.end(),
                   inserter(unmapped, unmapped.begin()));
    for (p &o_t: unmapped) {
        int m_dist = abs(o_t.first - pos.first) + abs(o_t.second - pos.second);
        if (m_dist > r_max) {
            continue;
        }
        return false;
    }
    return true;
}

int main() {
    int x, y;
    scanf("%d %d %d\n", &n_t, &n_s, &r_max);
    for (int i = 0; i < n_t; i++) {
        int t_x, t_y;
        scanf("%d %d\n", &x, &y);
        trees.insert(pair(t_x, t_y));
    }
    for (int i = 0; i < n_s; i++) {
        scanf("%d %d\n", &x, &y);
        sensors[i] = pair(x, y);
    }
    for (char d : {'R', 'U', 'L', 'D' }) {
        for (auto& t : trees) {
            p a = adj(sensors[0], (direction)(d)); // x = t_x + s_x
            p pos = pair(t.first - a.first, t.second - a.second); // position to check
            if (trees.find(pos) != trees.end()) {
                continue;
            }
            if (!check(pos, t, (direction)d)) {
                continue;
            }
            if (is_found) {
                cout << "Ambiguous" << endl;
            }
            found = pos;
        }
    }
    if (!is_found) {
        cout << "Impossible" << endl;
    }
    cout << found.first << " " << found.second << endl;
}

